libyul/YulName.cpp

```diff
@@ -0,0 +1,475 @@
+		m_predefined.sub = defineName("sub");
+
+	{
+		auto types = m_dialectTypes;
+		if (types.empty())
+			types.emplace_back(0, "");
+
+		for (auto const& [typeName, typeLabel]: types)
+		{
+			if (auto const* discardFunction = m_dialect.discardFunction(YulString(typeLabel)))
+				m_predefinedBuiltinFunctions.discardFunctions.emplace_back(nameOfBuiltin(discardFunction->name.str()));
+			else
+				m_predefinedBuiltinFunctions.discardFunctions.emplace_back(std::nullopt);
+
+			if (auto const* equalityFunction = m_dialect.equalityFunction(YulString(typeLabel)))
+				m_predefinedBuiltinFunctions.equalityFunctions.emplace_back(nameOfBuiltin(equalityFunction->name.str()));
+			else
+				m_predefinedBuiltinFunctions.equalityFunctions.emplace_back(std::nullopt);
+
+			if (auto const* booleanNegationFunction = m_dialect.booleanNegationFunction())
+				m_predefinedBuiltinFunctions.booleanNegationFunction = nameOfBuiltin(booleanNegationFunction->name.str());
+			else
+				m_predefinedBuiltinFunctions.booleanNegationFunction = std::nullopt;
+
+			if (auto const* memStoreFunction = m_dialect.memoryStoreFunction(YulString(typeLabel)))
+				m_predefinedBuiltinFunctions.memoryStoreFunctions.emplace_back(nameOfBuiltin(memStoreFunction->name.str()));
+			else
+				m_predefinedBuiltinFunctions.memoryStoreFunctions.emplace_back(std::nullopt);
+
+			if (auto const* memLoadFunction = m_dialect.memoryLoadFunction(YulString(typeLabel)))
+				m_predefinedBuiltinFunctions.memoryLoadFunctions.emplace_back(nameOfBuiltin(memLoadFunction->name.str()));
+			else
+				m_predefinedBuiltinFunctions.memoryLoadFunctions.emplace_back(std::nullopt);
+
+			if (auto const* storageStoreFunction = m_dialect.storageStoreFunction(YulString(typeLabel)))
+				m_predefinedBuiltinFunctions.storageStoreFunctions.emplace_back(nameOfBuiltin(storageStoreFunction->name.str()));
+			else
+				m_predefinedBuiltinFunctions.storageStoreFunctions.emplace_back(std::nullopt);
+
+			if (auto const* storageLoadFunction = m_dialect.storageLoadFunction(YulString(typeLabel)))
+				m_predefinedBuiltinFunctions.storageLoadFunctions.emplace_back(nameOfBuiltin(storageLoadFunction->name.str()));
+			else
+				m_predefinedBuiltinFunctions.storageLoadFunctions.emplace_back(std::nullopt);
+
+			m_predefinedBuiltinFunctions.hashFunctions.emplace_back(nameOfBuiltin(m_dialect.hashFunction(YulString(typeLabel)).str()));
+		}
+	}
+
+	m_predefined.placeholder_zero = defineName("@ 0");
+	m_predefined.placeholder_one = defineName("@ 1");
+	m_predefined.placeholder_thirtytwo = defineName("@ 32");
+}
+
+YulNameRepository::BuiltinFunction const* YulNameRepository::builtin(YulName const _name) const
+{
+	yulAssert(nameWithinBounds(_name), "YulName exceeds repository size, probably stems from another instance.");
+	auto const baseName = baseNameOf(_name);
+	if (isBuiltinName(baseName))
+	{
+		auto const it = m_builtinFunctions.find(_name);
+		if (it != m_builtinFunctions.end())
+			return &it->second;
+		else
+			return nullptr;
```

Comment:
I think the `else` branch is not necessary, since if it doesn't return in the if above it will reach the function return outside the `if` scope.
```suggestion
		return nullptr;
```

So the whole `if` block could be rewritten to something like:
```c++
if (isBuiltinName(baseNameOf(_name)))
{
	auto const it = m_builtinFunctions.find(_name);
	if (it != m_builtinFunctions.end())
		return &it->second;
}
return nullptr;
```